util/udev/qmk_id.c

```diff
@@ -0,0 +1,174 @@
+    *item = ntohl(r);
+    return 0;
+}
+
+// including bType and bTag
+enum hid_item_type {
+    HID_TYPE_USAGE_PAGE = 0x01,
+    HID_TYPE_USAGE      = 0x02,
+    HID_TYPE_COLLECTION = 0x28,
+    HID_TYPE_LONG_ITEM  = 0xFF,
+};
+
+enum hid_collection_type {
+    HID_COLLECTION_APPLICATION = 0x01,
+};
+
+static int parse_type(const uint8_t *report, size_t report_size, size_t *offset, enum hid_item_type *item_type, uint8_t *item_size) {
+    if (*offset >= report_size) {
+        return -EIO;
+    }
+
+    uint8_t data = report[*offset];
+
+    // long items
+    if (data == 0xFE) {
+        *item_type = HID_TYPE_LONG_ITEM;
+        if (*offset + 2 >= report_size) {
+            return -EIO;
+        }
+        *item_size = report[*offset + 1];
+        *offset    = *offset + 2;
+        return 0;
+    }
+
+    *item_type = (enum hid_item_type)((data & 0xFC) >> 2);
+    *item_size = data & 0x03;
+    if (*item_size == 3) {
+        *item_size = 4;
+    }
+    *offset = *offset + 1;
+    return 0;
+}
+
+static int is_collection(const uint8_t *report, size_t report_size, uint32_t usage, enum hid_collection_type collection_type) {
+    uint32_t up = 0, u = 0, ct = 0;
+    size_t   offset = 0;
+    int      ret;
+
+    while (1) {
+        uint8_t            item_size;
+        enum hid_item_type item_type;
+        ret = parse_type(report, report_size, &offset, &item_type, &item_size);
+        if (ret) {
+            return ret;
+        }
+        if (item_type == HID_TYPE_LONG_ITEM) {
+            offset += item_size;
+            continue;
+        }
+        uint32_t item;
+        ret = parse_item(report, report_size, &offset, item_size, &item);
+        if (ret) {
+            return ret;
+        }
+        switch (item_type) {
+            case HID_TYPE_USAGE_PAGE:
+                up = item;
+                break;
+            case HID_TYPE_USAGE:
+                if (item_size <= 2) {
+                    item = (up << 16) | item;
+                }
+                u = item;
+                break;
+            case HID_TYPE_COLLECTION:
+                ct = item;
+                break;
+            case HID_TYPE_LONG_ITEM:
+                // can never happen
+                abort();
+        }
+        if (u && ct) {
+            return usage == u && collection_type == ct;
+        }
+    }
+}
+
+static int is_teensy_style_console(const uint8_t *report, size_t report_size) { return is_collection(report, report_size, 0xFF310074, HID_COLLECTION_APPLICATION); }
+
+int main(int argc, char **argv) {
+    if (argc != 2) {
+        return EINVAL;
+    }
+
+    int _cleanup_close sysfs_fd = openat(-1, argv[1], O_RDONLY);
+    if (sysfs_fd == -1) {
+        return errno;
+    }
+    int _cleanup_close report_fd = openat(sysfs_fd, "device/report_descriptor", O_RDONLY);
+    if (report_fd == -1) {
+        return errno;
+    }
+
+    close_fd(&sysfs_fd);
+    close_fd(&report_fd);
+
+    if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_STRICT) == -1) {
```

Comment:
Unfortunately this does not work because the processes spawned by udev already are using seccomp filters.
seccomp filters and seccomp strict mode however are not compatible.
The alternative would be to add an additional custom filter that has the same effect but is composable with the existing filter.